ee/control/actionqueue/actionqueue.go

```diff
@@ -0,0 +1,241 @@
+package actionqueue
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/agent/storage/inmemory"
+	"github.com/kolide/launcher/pkg/agent/types"
+)
+
+const (
+	ActionsSubsystem       = "actions"
+	defaultCleanupInterval = time.Hour * 12
+	// about 6 months, long enough to ensure that K2 no longer has the message
+	// and will not send a duplicate
+	actionRetentionPeriod = time.Hour * 24 * 30 * 6
+)
+
+type actor interface {
+	Do(data io.Reader) error
+}
+
+type action struct {
+	ID          string    `json:"id"`
+	ValidUntil  int64     `json:"valid_until"` // timestamp
+	Type        string    `json:"type"`
+	ProcessedAt time.Time `json:"processed_at,omitempty"`
+}
+
+type actionqueue struct {
+	ctx                   context.Context
+	actors                map[string]actor
+	store                 types.KVStore
+	logger                log.Logger
+	actionCleanupInterval time.Duration
+	cancel                context.CancelFunc
+}
+
+type actionqueueOption func(*actionqueue)
+
+func WithLogger(logger log.Logger) actionqueueOption {
+	return func(aq *actionqueue) {
+		aq.logger = logger
```

Comment:
```suggestion
		aq.logger = log.With(logger, "component", "actionqueue")
```

I've found the `With component` we've been doing in other places super helpful for understanding and filtering logs -- could we include it here too? Sorry I didn't notice this on the last pass through!

Reply:
done